src/main/java/org/opensearch/security/action/apitokens/ApiTokenIndexHandler.java

```diff
@@ -0,0 +1,147 @@
+import org.opensearch.core.xcontent.DeprecationHandler;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.ToXContent;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.index.reindex.BulkByScrollResponse;
+import org.opensearch.index.reindex.DeleteByQueryAction;
+import org.opensearch.index.reindex.DeleteByQueryRequest;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.security.support.ConfigConstants;
+
+public class ApiTokenIndexHandler {
+
+    private Client client;
+    private ClusterService clusterService;
+    private static final Logger LOGGER = LogManager.getLogger(ApiTokenIndexHandler.class);
+
+    public ApiTokenIndexHandler(Client client, ClusterService clusterService) {
+        this.client = client;
+        this.clusterService = clusterService;
+    }
+
+    public String indexToken(ApiToken token) {
+        try (final ThreadContext.StoredContext ctx = client.threadPool().getThreadContext().stashContext()) {
+
+            XContentBuilder builder = XContentFactory.jsonBuilder();
+            String jsonString = token.toXContent(builder, ToXContent.EMPTY_PARAMS).toString();
+
+            IndexRequest request = new IndexRequest(ConfigConstants.OPENSEARCH_API_TOKENS_INDEX).source(jsonString, XContentType.JSON);
+
+            ActionListener<IndexResponse> irListener = ActionListener.wrap(idxResponse -> {
+                LOGGER.info("Created {} entry.", ConfigConstants.OPENSEARCH_API_TOKENS_INDEX);
+            }, (failResponse) -> {
+                LOGGER.error(failResponse.getMessage());
+                LOGGER.info("Failed to create {} entry.", ConfigConstants.OPENSEARCH_API_TOKENS_INDEX);
+            });
+
+            client.index(request, irListener);
+            return token.getDescription();
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    public void deleteToken(String name) throws ApiTokenException {
+        try (final ThreadContext.StoredContext ctx = client.threadPool().getThreadContext().stashContext()) {
+            DeleteByQueryRequest request = new DeleteByQueryRequest(ConfigConstants.OPENSEARCH_API_TOKENS_INDEX).setQuery(
+                QueryBuilders.matchQuery("description", name)
+            ).setRefresh(true);  // This will refresh the index after deletion
+
+            BulkByScrollResponse response = client.execute(DeleteByQueryAction.INSTANCE, request).actionGet();
+
+            long deletedDocs = response.getDeleted();
+
+            if (deletedDocs == 0) {
+                throw new ApiTokenException("No token found with name " + name);
+            }
+            LOGGER.info("Deleted " + deletedDocs + " documents");
+        }
+    }
+
+    public Map<String, ApiToken> getApiTokens() {
```

Comment:
Should we consider building filtering into this. i.e. passing `name` as a param to filter the results?

Reply:
I just wanted to limit the code for now. In the future this may make sense, but I also don't expect users to create too many tokens to begin with, so not sure if this will make the cut or not, so for now I left it out.